From a5aaa426ad49246015e2b995ffa2ba9f9f9a791f Mon Sep 17 00:00:00 2001
From: Simon McVittie <simon.mcvittie@collabora.co.uk>
Date: Tue, 17 Dec 2013 16:54:48 +0000
Subject: Include ddebs in reprepro too, and fix inclusion of udebs

We previously looked for Package-Type: in the .changes file to
decide whether it included udebs, but that doesn't work: Package-Type
only appears in the binary package control data (DEBIAN/control).
Parse the binary packages' filenames instead.

Bug: Collabora #397
Reviewed-by: Sjoerd Simons <sjoerd.simons@collabora.co.uk>

--- a/src/backend/bs_publish
+++ b/src/backend/bs_publish
@@ -627,21 +627,42 @@
   for my $f (@changed) {
       if ($f =~ /\.changes/) {
         print " Updated changes file => $f\n" ;
-        qsystem ('reprepro', '-b', $repo,
-          '--ignore=wrongdistribution',
-          '--ignore=updatedarchall',
-	  '--ignore=unusedarch',
-	  '-T', 'deb',
-          '-C', $component,
-          'include', $codename,
-          "$extrep/$f");
-        if (qsystem ('grep', '-q', '^Package-Type:.*',
-            "$extrep/$f") == 0) {
+        my %types;
+        if (open (my $fh, '<', "$extrep/$f")) {
+          # Read the .changes file looking for binary packages
+          # (.deb, .udeb, .ddeb)
+          my $in_files = 0;
+          while (my $line = <$fh>) {
+            if ($in_files) {
+              if ($line =~ /^\s/) {
+                if ($line =~ /\.((?:|u|d)deb)$/) {
+                  $types{$1} = 1;
+                }
+              } else {
+                $in_files = 0;
+              }
+            } elsif ($line =~ /^Files\s*:/) {
+              $in_files = 1;
+            }
+          }
+        } else {
+          print "    unable to open $extrep/$f: $!\n";
+        }
+
+        if (!%types) {
+          # no binary packages at all? assume we must have mis-parsed it,
+          # and run reprepro for the .deb files so we get a better
+          # error report
+          %types = (deb => 1);
+        }
+
+        foreach my $type (keys %types) {
+
           qsystem ('reprepro', '-b', $repo,
             '--ignore=wrongdistribution',
             '--ignore=updatedarchall',
             '--ignore=unusedarch',
-            '-T', 'udeb',
+            '-T', $type,
             '-C', $component,
             'include', $codename,
             "$extrep/$f");
